File: 0646-maximum-length-of-pair-chain/0646-maximum-length-of-pair-chain.cpp
class Solution {
public:
    vector<int>memo;
    int findLongestChain(vector<vector<int>>& pairs) {
        sort(pairs.begin(), pairs.end(),
             [](const vector<int>& a, const vector<int>& b) {
                 return a[1] < b[1];
             });
        int n = pairs.size();
        memo = vector<int>(n, 0);
        return solve(pairs, 0, INT_MIN);
    }

    int solve(vector<vector<int>>& pairs, int index, int prev) {
        if (index == pairs.size()) {
            return 0;
        }
        if(memo[index]){
            return memo[index];
        }
        int take = 0, skip = 0;
        if (prev < pairs[index][0]) {
            take = 1 + solve(pairs, index + 1, pairs[index][1]);
        }
        skip = solve(pairs, index + 1, prev);

        return memo[index] = max(take, skip);
    }
};

File: 0646-maximum-length-of-pair-chain/0646-maximum-length-of-pair-chain.py
class Solution:
    def findLongestChain(self, pairs: List[List[int]]) -> int:
        n = len(pairs)
        pairs.sort()
        # print(pairs)
        memo = [[-1] * n for _ in range(n)]

        def solve(index, prev):
            if index == n:
                return 0

            if memo[index][prev] != -1:
                return memo[index][prev]

            take, skip = 0, 0
            if prev == -1 or pairs[index][0] > pairs[prev][1]:
                take = 1 + solve(index + 1, index)
            skip = solve(index + 1, prev)
            # print("take: ", take)
            # print("skip: ", skip)
            # print(pairs[index])
            memo[index][prev] = max(take, skip)
            return memo[index][prev]

        return solve(0, -1)
